src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java

```diff
@@ -36,13 +38,32 @@ public UUID send(LetterRequest letter, String serviceName) {
 
         final String messageId = generateChecksum(letter);
 
-        log.info("Generated message: id = {} for letter with print queue id = {}", messageId, letter.type);
+        log.info("Generated message: id = {}", messageId);
 
-        byte[] pdf = pdfCreator.create(letter);
-        Letter dbLetter = new Letter(messageId, serviceName, null, letter.type, pdf);
+        Optional<Letter> duplicateLetter = letterRepository.findByMessageIdAndStateOrderByCreatedAtDesc(
+            messageId,
+            LetterState.Created
+        );
+
+        return duplicateLetter
+            .map(l -> {
+                UUID id = l.getId();
+                log.info("Same message found already created. Returning letter id {} instead", id);
+                return id;
+            })
+            .orElseGet(() -> saveNewLetterAndReturnId(letter, messageId, serviceName));
+    }
+
+    private UUID saveNewLetterAndReturnId(LetterRequest letterRequest, String messageId, String serviceName) {
+        byte[] pdf = pdfCreator.create(letterRequest);
+
+        Letter letter = new Letter(messageId, serviceName, null, letterRequest.type, pdf);
```

Comment:
why `null`?

Reply:
additional data is not needed any more since we generate pdf in advance

Reply:
it is still needed